### PR TITLE
feat: Using gRPC Bidirectional Streaming for Map

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: pkg/mapper/service.go [hunk 1]
```diff
@@ -33,29 +36,126 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*mappb.ReadyRespons
...
+	responseCh := make(chan *mappb.MapResponse, 500) // FIXME: identify the right buffer size
+	defer close(responseCh)
+
+	// Dedicated goroutine to send responses to the stream
+	g.Go(func() error {
+		for {
+			select {
+			case resp := <-responseCh:
+				if err := stream.Send(resp); err != nil {
+					log.Printf("failed to send response: %v", err)
```
Comment (BulkBeing) [inline pkg/mapper/service.go:62]:
```suggestion
					log.Printf("Failed to send response: %v", err)
```

File: pkg/mapper/service.go [hunk 2]
```diff
@@ -33,29 +36,126 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*mappb.ReadyRespons
...
+		select {
+		case <-grpCtx.Done():
+			break outer
+		default:
+			req, err := stream.Recv()
+			if err == io.EOF {
+				break outer
+			}
+			if err != nil {
+				log.Printf("failed to receive request: %v", err)
```
Comment (BulkBeing) [inline pkg/mapper/service.go:84]:
```suggestion
				log.Printf("Failed to receive request: %v", err)
```

File: pkg/mapper/service.go [hunk 3]
```diff
@@ -33,29 +36,126 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*mappb.ReadyRespons
...
+	})
+
+	var readErr error
+	// Read requests from the stream and process them
+outer:
+	for {
+		select {
+		case <-grpCtx.Done():
+			break outer
+		default:
```
Comment (BulkBeing) [inline pkg/mapper/service.go:78]: one level indentation can be removed by empty `default` case.
```go
select {
case <-grpCtx.Done():
	break outer
default:
}
```

File: pkg/mapper/service_test.go [hunk 4]
```diff
@@ -2,118 +2,292 @@ package mapper
...
 		name    string
 		handler Mapper
 		args    args
-		want    *mappb.MapResponse
-		wantErr bool
+		want    *proto.MapResponse
 	}{
 		{
-			name: "map_fn_forward_msg",
+			name: "sourceTransform_fn_forward_msg",
```
Comment (BulkBeing) [inline pkg/mapper/service_test.go:83]:
```suggestion
			name: "map_fn_forward_msg",
```

File: pkg/mapper/service_test.go [hunk 5]
```diff
@@ -2,118 +2,292 @@ package mapper
...
 						Keys:  []string{"client_test"},
 						Value: []byte(`test`),
 					},
 				},
 			},
-			wantErr: false,
 		},
 		{
-			name: "map_fn_forward_msg_forward_to_all",
+			name: "sourceTransform_fn_forward_msg_forward_to_all",
```
Comment (BulkBeing) [inline pkg/mapper/service_test.go:109]:
```suggestion
			name: "map_fn_forward_msg_forward_to_all",
```

File: pkg/mapper/service_test.go [hunk 6]
```diff
@@ -2,118 +2,292 @@ package mapper
...
 					{
 						Value: []byte(`test`),
 					},
 				},
 			},
-			wantErr: false,
 		},
 		{
-			name: "map_fn_forward_msg_drop_msg",
+			name: "sourceTransform_fn_forward_msg_drop_msg",
```
Comment (BulkBeing) [inline pkg/mapper/service_test.go:134]:
```suggestion
			name: "map_fn_forward_msg_drop_msg",
```

**Review by BulkBeing — APPROVED** [review]

**Review by vigith — APPROVED** [review]

**Pull request merged**